### ISSUE TITLE
backend_type in layer class constructor

### ISSUE THREAD

**Issue by beru** [tiny-dnn/tiny-dnn#550]
I see `backend_type` parameter in some layer classes constructors such as `convolutional_layer`, `deconvolutional_layer`, `fully_connected_layer`, `max_pooling_layer`, `quantized_convolutional_layer`, `quantized_deconvolutional_layer`, `quantized_fully_connected_layer`.

I'm not sure why it needs to be specified at layer level. I just thought setting a backend at network level is much easier for users.

I probablly asked about the reason before but completely lost the memory...

**Comment by edgarriba** [comment]
@beru the common use case might be the one you point out, a single backend per network. However, as @bhack comment, some backends do not cover all the operations. Also might be the case that for whatever reason you want to experiment with different backends. Finally, debugging was another reason which for I decided to implement this design.

Notice that there might some confusing variable names like backend/engine. This is still here due to the unfinished refactoring of old backend class (quantized kernels are blocking to finish this task).

**Comment by bhack** [comment]
And remember that changing engine node by node could require Tensor memory copying overhead, scheduling problems etc. So this will be a major problem of op placement.

**Comment by edgarriba** [comment]
right, but we could have a high level routine to set a single backend per network. I'd like to keep backend type per layer in order to have control over testing.

**Comment by beru** [comment]
@bhack @edgarriba Thank you for the explanation.

I think default backend for layers should be able to be set at network level since it lowers coding cost.

And if some users want to change backend type for specific layers, it would be good if they can change it by calling layer class method.

One problem I can think of by setting backend per layer level is, in case of quantized convolutional neural networks, some layers may not be able to use quantized data if the quantized data is used network-wide.
(It isn't a problem right now as float_t conversion takes place inside kernel but doing so slows down operation)

Keeping backend type per layer may not be that bad as it gives us flexibility in network construction, but some kernels such as `FullyConnectedGradOp` do not ready fallback paths when certain backend does not support its operation.

**Comment by beru** [comment]
@edgarriba Sure. This one is a trivial matter.
